org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java

```diff
@@ -39,10 +42,28 @@ public TreeViewerDoubleClickListener(String commandId, TreeViewer treeViewer) {
 
     @Override
     public void doubleClick(DoubleClickEvent event) {
-        NodeSelection selection = NodeSelection.from(this.treeViewer.getSelection());
-        if (isEnabledFor(selection)) {
-            run();
+        final IStructuredSelection structuredSelection = (IStructuredSelection)event.getSelection();
+        if (structuredSelection == null || structuredSelection.isEmpty()) {
+            return;
         }
+
+        final Object selection = structuredSelection.getFirstElement();
+
+        final ITreeContentProvider provider = (ITreeContentProvider)treeViewer.getContentProvider();
+
+        if (provider.hasChildren(selection)) {
+            if (treeViewer.getExpandedState(selection)) {
+                treeViewer.collapseToLevel(selection, AbstractTreeViewer.ALL_LEVELS);
+            }
+            else {
+                treeViewer.expandToLevel(selection, 1);
+            }
+        } else {
+	        NodeSelection nodeSelection = NodeSelection.from(this.treeViewer.getSelection());
+	        if (isEnabledFor(nodeSelection)) {
+	            run();
+	        }
+	    }
     }
 
     private boolean isEnabledFor(NodeSelection node) {
```

Comment:
After discussing it internally we concluded that we should disable running the default tasks by double-clicking on the project tasks in favor of expanding-collapsing the node. To implement this, `isEnabledFor(NodeSelection)` should return `TaskViewActionStateRules.taskScopedTaskExecutionActionsEnablement(selection).asBoolean()`.